compiler/rustc_mir_transform/src/dest_prop.rs

```diff
@@ -353,524 +336,564 @@ impl<'tcx> MutVisitor<'tcx> for Replacer<'tcx> {
+            true
         }
     }
+}
 
-    fn record_local_conflict(&mut self, a: Local, b: Local, why: &str) {
-        trace!("conflict {:?} <-> {:?} due to {}", a, b, why);
-        self.matrix.insert(a, b);
-        self.matrix.insert(b, a);
-    }
+/// Describes where a statement/terminator writes to and reads from.
+///
+/// We actually also use this data for an additional performance optimization: The locals that
+/// appear in these lists are the full set of all locals that might have their liveness data change
+/// as a result of the statement/terminator. This means that when we update dataflow conflicts next
+/// time, we only need to consider candidates involving locals that appear here. This prevents us
+/// from having to look at the whole candidate set each iteration.
+#[derive(Default, Debug)]
+struct InteriorBehavior {
+    reads: Vec<Local>,
+    writes: Vec<Local>,
+}
 
-    /// Records locals that must not overlap during the evaluation of `stmt`. These locals conflict
-    /// and must not be merged.
-    fn record_statement_conflicts(&mut self, stmt: &Statement<'_>) {
-        match &stmt.kind {
-            // While the left and right sides of an assignment must not overlap, we do not mark
-            // conflicts here as that would make this optimization useless. When we optimize, we
-            // eliminate the resulting self-assignments automatically.
-            StatementKind::Assign(_) => {}
-
-            StatementKind::SetDiscriminant { .. }
-            | StatementKind::Deinit(..)
-            | StatementKind::StorageLive(..)
-            | StatementKind::StorageDead(..)
-            | StatementKind::Retag(..)
-            | StatementKind::FakeRead(..)
-            | StatementKind::AscribeUserType(..)
-            | StatementKind::Coverage(..)
-            | StatementKind::CopyNonOverlapping(..)
-            | StatementKind::Nop => {}
-        }
+impl InteriorBehavior {
+    fn reset(&mut self) {
+        self.reads.clear();
+        self.writes.clear();
     }
 
-    fn record_terminator_conflicts(&mut self, term: &Terminator<'_>) {
-        match &term.kind {
-            TerminatorKind::DropAndReplace {
-                place: dropped_place,
-                value,
-                target: _,
-                unwind: _,
-            } => {
-                if let Some(place) = value.place()
-                    && !place.is_indirect()
-                    && !dropped_place.is_indirect()
-                {
-                    self.record_local_conflict(
-                        place.local,
-                        dropped_place.local,
-                        "DropAndReplace operand overlap",
-                    );
+    fn for_statement<'tcx>(&mut self, statement: &StatementKind<'tcx>) {
+        self.reset();
+        match statement {
+            StatementKind::Assign(assign) => {
+                let lhs = assign.0;
+                self.add_write(lhs);
+                let rvalue = &assign.1;
```

Comment:
`StatementKind::Assign(box (lhs, rvalue)) => ...`

Reply:
I can change this if you feel strongly in that direction, but I intentionally avoided doing this because box patterns are a nightly feature and one that is not on the path to stabilization. Since presumably they'll eventually be removed, I don't really want to add any more code that uses them.

Reply:
I would expect that box patterns would be migrated to deref patterns, so I wouldn't worry about that.